src/python/pants/backend/python/macros/deprecation_fixers.py

```diff
@@ -145,6 +159,86 @@ async def determine_macro_changes(all_targets: AllTargets, _: MacroRenamesReques
     return MacroRenames(tuple(sorted(generators)), FrozenDict(sorted(generated.items())))
 
 
+def maybe_address(val: str, renames: MacroRenames, *, relative_to: str | None) -> Address | None:
+    # All macros generate targets with a `name`, so we know they must have `:`. We know they
+    # also can't have `#` because they're not generated targets syntax.
+    if ":" not in val or "#" in val:
+        return None
+
+    try:
+        # We assume that all addresses are normal addresses, rather than file addresses, as
+        # we know that none of the generated targets will be file addresses. That is, we can
+        # ignore file addresses.
+        addr = AddressInput.parse(val, relative_to=relative_to).dir_to_address()
+    except InvalidAddress:
+        return None
+
+    return addr if addr in renames.generated else None
+
+
+def new_addr_spec(original_val: str, new_addr: Address) -> str:
+    # Preserve relative addresses (`:tgt`), else use the normalized spec.
+    if original_val.startswith(":"):
+        return (
+            f"#{new_addr.generated_name}"
+            if new_addr.is_default_target
+            else f":{new_addr.target_name}#{new_addr.generated_name}"
+        )
+    return new_addr.spec
+
+
+class OptionsChecker:
+    """Checks a hardcoded list of options for using deprecated addresses.
+
+    This is returned as a rule so that it acts as a singleton.
+    """
+
+
+class OptionsCheckerRequest:
+    pass
+
+
+@rule(desc="Check option values for Python macro syntax vs. target generator", level=LogLevel.DEBUG)
+async def maybe_warn_options_macro_references(
+    _: OptionsCheckerRequest,
+    python_protobuf: PythonProtobufSubsystem,
+    python_thrift: ThriftPythonSubsystem,
+    flake8: Flake8,
+    pylint: Pylint,
+    mypy: MyPy,
+) -> OptionsChecker:
+    renames = await Get(MacroRenames, MacroRenamesRequest())
+
+    opt_to_renames: DefaultDict[str, set[tuple[str, str]]] = defaultdict(set)
+
+    def check(deps: UnparsedAddressInputs, option: str) -> None:
+        for runtime_dep in deps.values:
+            addr = maybe_address(runtime_dep, renames, relative_to=None)
+            if addr:
+                new_addr = new_addr_spec(runtime_dep, renames.generated[addr][0])
+                opt_to_renames[option].add((runtime_dep, new_addr))
+
+    check(python_protobuf.runtime_dependencies, "[python-protobuf].runtime_dependencies")
+    check(python_thrift.runtime_dependencies, "[python-thrift].runtime_dependencies")
+    check(flake8.source_plugins, "[flake8].source_plugins")
+    check(pylint.source_plugins, "[pylint].source_plugins")
+    check(mypy.source_plugins, "[mypy].source_plugins")
```

Comment:
It's unlikely that these options will have bad references, but it's cheap for us to check.